Nonprofit-Exchange-Hub/web-app
# Esteban/poc/socket io/messages

## Dev Summary
The purpose of this summary is to show what it takes to bootstrap websockets/polling with auth in the back-end and client.
- makes use of `socket-io`'s socket implementation, which uses both polling and a ws connection.
- Adds a guard that works with a ws request object (it's different than an http request object)
- Adds a rudimentary ui for chat on a protected ui route
- I added all new resource (BE) and feature route (FE) so that it's easy to remove later




## Test Plan


repro steps:

1. run the app BE & FE
2. You should create 2 users in your local env.
3. open 2 incognito browser windows.
4. Log into each user in each incognito window
5. got to `http://localhost:3000/chat`
expected: 


Expect to
- see prompt when the other user is typing
- See messages come in real-time
- Messages will save to the database
- Each message has the users's first name under the name field


## Resources

- [general guide on sockets / `NestJS`](https://gabrieltanner.org/blog/nestjs-realtime-chat)
- [`NestJS` Guards in sockets docs](https://docs.nestjs.com/websockets/guards#binding-guards)
- [on `websockets context`](https://docs.nestjs.com/fundamentals/execution-context#host-handler-arguments)
- [how to get cookies from the ws request object](https://github.com/nestjs/nest/issues/1254#issuecomment-453280476)
- [Adding user to the request](https://stackoverflow.com/a/61815880/13862254)
